src/models/ae.jl

```diff
@@ -0,0 +1,160 @@
+struct AE
+	encoder
+	decoder
+end
+
+Flux.@functor AE
+
+function (ae::AE)(x)
+	z = ae.encoder(x)
+	x̂ = ae.decoder(z)
+	return x̂
+end
+
+# ae_loss already in svdd
+
+function conv_ae_constructor(
+	;idim=(2,2,1), 
+	zdim::Int=1, 
+	activation="relu", 
+	kernelsizes=(1,1), 
+	channels=(1,1), 
+	scalings=(1,1),
+	init_seed=nothing, 
+	batchnorm=false, 
+	kwargs...
+)
+	# if seed is given, set it
+	(init_seed !== nothing) ? Random.seed!(init_seed) : nothing
+	# encoder 
+	encoder = Chain(
+		conv_encoder(idim, zdim, kernelsizes, channels, scalings; activation=activation, batchnorm=batchnorm)...
+	)
+	# decoder 
+	decoder = Chain(
+		conv_decoder(idim, zdim, reverse(kernelsizes), reverse(channels), reverse(scalings),
+			activation=activation, vec_output=false, vec_output_dim=nothing, batchnorm=batchnorm)...,
+			x->sigmoid.(x) # our dataset images are scaled from 0-1
+	)
+	# reset seed
+	(init_seed !== nothing) ? Random.seed!() : nothing
+
+	model = AE(encoder, decoder)
+end
+
+
+function ae_constructor(
+	;idim::Int=1, 
+	zdim::Int=1, 
+	hdim::Int=64, 
+	activation = "relu", 
+	nlayers::Int=3, 
+	init_seed=nothing, 
+	kwargs...
+	)
+	(init_seed !== nothing) ? Random.seed!(init_seed) : nothing
+
+	encoder = build_mlp(idim, hdim, zdim, nlayers, activation=activation, lastlayer="linear")
+
+	decoder = build_mlp(zdim, hdim, idim, nlayers, activation=activation, lastlayer="linear")
+
+	(init_seed !== nothing) ? Random.seed!() : nothing
+
+	return AE(encoder, decoder)
+end
+
+function anomaly_score(ae::AE, real_input; dims=(1,2,3), to_testmode::Bool=true)
+	(to_testmode == true) ? Flux.testmode!(ae) : nothing
+	score = vec(Flux.sum((ae(real_input) .- real_input).^2, dims=dims))
+	(to_testmode == true) ? Flux.testmode!(ae, false) : nothing
+	return score
+end
+
+function anomaly_score_gpu(ae::AE, real_input; batch_size=64, dims=(1,2,3), to_testmode::Bool=true)
+	real_input = Flux.Data.DataLoader(real_input, batchsize=batch_size)
+	(to_testmode == true) ? Flux.testmode!(ae) : nothing
+	ae = ae |> gpu
+	scores = Array{Float32}([])
+	for x in real_input
+		x = x |> gpu
+		score = vec(Flux.sum((ae(x) .- x).^2, dims=dims))
+		scores = cat(scores, score |> cpu, dims=1)
+	end
+	(to_testmode == true) ? Flux.testmode!(ae, false) : nothing
+	return scores
+end
+
+
+function StatsBase.fit!(ae::AE, data, params)
+	# prepare batches & loaders
+	train_loader, val_loader = prepare_dataloaders(data, params)
```

Comment:
What is this?

Reply:
you mean `train_loader, val_loader = prepare_dataloaders(data, params)`?

It is just function that create dataloaders for training 

```julia
function prepare_dataloaders(data; batch_size::Int=64, iters::Int=10000)
	train_loader = MLDataPattern.RandomBatches(data[1][1], size=batch_size, count=iters)
	# for cheching convergence I need to drop anomal samples from validation data
	val_data_ind = findall(x->x==0, data[2][2])
	if length(size(data[2][1]))==4
		val_data = data[2][1][:,:,:,val_data_ind]
	else
		val_data = data[2][1][:,val_data_ind]
	end
	val_loader = Flux.Data.DataLoader(val_data, batchsize=batch_size)
	return train_loader, val_loader
end
```